### DISCUSSION TITLE
Argo CD UI is slow with 1.5k+ applications

### DISCUSSION THREAD

**Discussion by rumstead** [argoproj/argo-cd#8446] [Q&A]
Hi all,

I am wondering how everyone tunes their Argo CD deploy to support having 1.5k+ applications. I have found that the UI is pretty slow to initially load and do any filtering. 

I run 2 argocd controllers with `controller.status.processors: "100"` and `controller.operation.processors: "50"`. I do have a little bit of CPU throttling, though we haven't seen any issues with slow syncs. Additionally, I have bumped `ARGOCD_GRPC_MAX_SIZE_MB` to 300 for the argocd server. 

Happy to contribute some docs around tuning. 

Also added gzip to the argocd server - https://argo-cd.readthedocs.io/en/stable/operator-manual/server-commands/argocd-server/#argocd-server. However since 2.8.0 this looks [to default true](https://github.com/argoproj/argo-cd/commit/a400ad11d7a66d87962faf1dcca6c305125c8879).

**Comment by rumstead** [comment]
I am compiling some of the data from the slack thread here.

Relevant Open Issues:
1. https://github.com/argoproj/argo-cd/issues/6125
2. https://github.com/argoproj/argo-cd/issues/7232
3. https://github.com/argoproj/argo-cd/issues/8175
4. https://github.com/argoproj/argo-cd/issues/9575

Relevant Closed Issues:
1. https://github.com/argoproj/argo-cd/issues/4226 
2. https://github.com/argoproj/argo-cd/issues/4225

**Comment by jcmcken** [comment]
I don't know much about the internals, but ultimately, I don't really think this is a tuning question necessarily but a question of filtering and paginating at the server instead of at the client. If the default 'per page' is 10, then ideally the server should query the data for only those 10 and send only those 10 to the browser (instead of querying for all and filtering clientside)

**Comment by tken2039** [comment]
Argo CD retrieves all Applications and then filters them based on user permissions (RBAC Policy). This filtering process has been identified as a cause of performance degradation. (Issue link: https://github.com/argoproj/argo-cd/issues/12985)

However, a modification to the filtering process has been made in [this PR](https://github.com/argoproj/argo-cd/issues/12985), and as of now, there exists a provisional solution as follows:

- By utilizing [UI Extensions](https://argo-cd.readthedocs.io/en/stable/developer-guide/extensions/ui-extensions/), users can enforce Project filters to expedite the retrieval of Applications.
  - To elaborate further, through UI Extensions, alter the value of `projectFilter` in the user's browser's local storage. This addition of a Project-based filtering query to API requests executed in the browser reduces the number of filtering processes based on permissions.

On the other hand, I'm delving into the code for a deeper understanding. The [casbin](https://github.com/casbin/casbin) package handles RBAC permission checks and relies on the [govaluate](https://github.com/Knetic/govaluate) package. My investigations suggest that the govaluate package's processes are slow. The casbin development team is actively addressing this performance issue, and I'm also exploring potential solutions.
(Issue link: https://github.com/casbin/casbin/issues/1168)

I would greatly appreciate if this information is of assistance to all.

**Comment by crenshaw-dev** [comment]
Server-side pagination will help a lot: https://github.com/argoproj/argo-cd/issues/14947
